frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/traverse/TraverseRepositoryAction.java

```diff
@@ -0,0 +1,283 @@
+      AnActionEvent anActionEvent) {
+    SyncToRemoteStatus status = gitMacheteBranch.getRelationToRemote().getSyncToRemoteStatus();
+    switch (status) {
+      case AheadOfRemote :
+        boolean pushApproved = true;
+        if (PropertiesComponent.getInstance().getBoolean(SHOW_PUSH_APPROVAL, /* defaultValue */ true)) {
+          val pushApprovalDialogBuilder = MessageDialogBuilder.okCancel(
+              getString("action.GitMachete.TraverseRepositoryAction.dialog.push-verification.title"),
+              getString("action.GitMachete.TraverseRepositoryAction.dialog.push-verification.text.HTML")
+                  .format(gitMacheteBranch.getName(), remoteTrackingBranch.getFullName()));
+
+          pushApprovalDialogBuilder
+              .yesText(getString("action.GitMachete.TraverseRepositoryAction.dialog.push-verification.ok-text"))
+              .icon(Messages.getQuestionIcon()).doNotAsk(new PushApprovalDialog());
+
+          pushApproved = pushApprovalDialogBuilder.ask(getProject(anActionEvent));
+        }
+
+        if (pushApproved) {
+          val fastForwardPushAction = ActionManager.getInstance()
+              .getAction(actionIdFormatString.format(PushCurrentAction.class.getSimpleName()));
+          fastForwardPushAction.actionPerformed(anActionEvent);
+        }
+        break;
+      case DivergedFromAndOlderThanRemote :
+      case DivergedFromAndNewerThanRemote :
+        val selectedAction = new DivergedFromRemoteDialog(getProject(anActionEvent), remoteTrackingBranch, gitMacheteBranch,
+            status).showAndGetThePreferredAction();
+        if (selectedAction == null) {
+          log().debug(
+              "Action selected for resolving divergence from remote is null: most likely the action has been canceled from Diverge-From-Remote-Dialog dialog");
+          return;
+        }
+        switch (selectedAction) {
+          case REBASE_ON_REMOTE :
+            break;
+          case MERGE_REMOTE_INTO_LOCAL :
+            break;
+          case RESET_ON_REMOTE :
+            val resetCurrentToRemoteAction = ActionManager.getInstance()
+                .getAction(actionIdFormatString.format(ResetCurrentToRemoteAction.class.getSimpleName()));
+            resetCurrentToRemoteAction.actionPerformed(anActionEvent);
+            break;
+          case FORCE_PUSH :
+            val pushAction = ActionManager.getInstance()
+                .getAction(actionIdFormatString.format(PushCurrentAction.class.getSimpleName()));
+            pushAction.actionPerformed(anActionEvent);
+            break;
+          default :
+            break;
+        }
+
+        break;
+
+      case BehindRemote :
+        boolean pullApproved = true;
+        if (PropertiesComponent.getInstance().getBoolean(SHOW_PULL_APPROVAL, /* defaultValue */ true)) {
+          val pullApprovalDialogBuilder = MessageDialogBuilder.okCancel(
```

Comment:
and here too, most it would be nice to bring some parts directly to PullApprovalDialog

Reply:
elaborations, pls 👍